### PR TITLE
Fix: fix Sudoku card fade reactivity for cross-suit updates(issue #3082)

### PR DESCRIPTION
**Pull request by panfs**
**Why bug occurs and how to test this fix**
- In previous logic, the appearance(faded or not) of color X is changed only when one X-color card is played or discarded. This logic is acceptable in most varients, but it may encounter rare edge cases shown below.
- We assume that there is a 4-player sudoku-varient game, and red starts with 3, yellow starts with 2, and blue starts with 5. There is a p5 held by some player. Now green and purple can begin with either 1 or 4.
- Because of some mistakes, all p2 are discarded. Now p5 still can be played if purple begins with 4.
- In another step, green begins with 4. According to sudoku's rule, purple can ONLY begins with 1. And because ALL p2 are discarded. p5 cannot be played anymore.
- After the analysis, we can find that due to sudoku's specific rule, the previous logic cannot handle this edge cases.
- To test this fix, you can simulate this edge case shown above.

**Commit description**
- `Sudoku: recompute cardStatus for ALL suits/ranks after play/discard to reflect cross-suit stack-start constraints.`
- `Handle scrubbed play actions by deriving suitIndex from the updated deck when available so playStackStarts/directions update reliably.`
- `Client: compute per-card status subscription using the same identity source as rendering (morphed → deck identity → cardIdentities), so trash/critical fades update immediately without requiring extra actions.`

### CODE REVIEW

**Comment by Zamiell** [comment]
can you add a test

File: packages/game/src/reducers/gameReducer.ts [hunk 1]
```diff
@@ -530,16 +541,33 @@ function gameReducerFunction(
...
-      gameState.cardStatus[action.suitIndex][rank] = getCardStatus(
-        action.suitIndex,
-        rank,
-        gameState.deck,
-        gameState.playStacks,
-        gameState.playStackDirections,
-        gameState.playStackStarts,
-        variant,
-      );
+    if (variant.sudoku) {
```
Comment (Zamiell) [inline packages/game/src/reducers/gameReducer.ts]: eh i dont like the variant specific code.
can you just simplify it to look through the whole deck, regardless of variant?
the perf wont matter afaik

File: packages/game/src/reducers/gameReducer.ts [hunk 2]
```diff
@@ -487,40 +488,50 @@ function gameReducerFunction(
...
-    );
-
-    gameState.playStackStarts[action.suitIndex] = getStackStartRank(
-      playStack,
-      gameState.deck,
-    );
+  if (action.type === "play" && variant.sudoku) {
+    const suitIndex =
+      action.suitIndex === -1
+        ? (gameState.deck[action.order]?.suitIndex ?? null)
```
Comment (Zamiell) [inline packages/game/src/reducers/gameReducer.ts]: use undefined instead of null please

File: packages/game/src/reducers/gameReducer.ts (hunk 1 above)
Comment (panfs) [reply packages/game/src/reducers/gameReducer.ts]: I've revised it.

File: packages/game/src/reducers/gameReducer.ts [hunk 3]
```diff
@@ -530,16 +541,19 @@ function gameReducerFunction(
...
-      gameState.cardStatus[action.suitIndex][rank] = getCardStatus(
-        action.suitIndex,
-        rank,
-        gameState.deck,
-        gameState.playStacks,
-        gameState.playStackDirections,
-        gameState.playStackStarts,
-        variant,
-      );
+    for (const i of variant.suits.keys()) {
```
Comment (Zamiell) [inline packages/game/src/reducers/gameReducer.ts:547]: can you put a code comment saying:
```
For a no-variant game, we would only have to iterate over cards of the same suit to check for non-playbility. But other variants become more complicated, so we iterate over the entire deck to be safe.
```

Comment (panfs) [reply packages/game/src/reducers/gameReducer.ts:547]: alright.

**Comment by Zamiell** [comment]
thanks. did you playtest this change on your local server to ensure that it works as expected?

**Comment by panfs** [comment]
Yes, it works as expected.

**Before y4 is played:**

截图 2026-04-14 00-56-43

---

**After y4 is played:**

截图 2026-04-14 00-57-28

**Comment by Zamiell** [comment]
excellent, thank you!
